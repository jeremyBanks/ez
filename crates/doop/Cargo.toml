cargo-features = ["workspace-inheritance"]

[package]
name = "doop"
description = "a loop-style macro for local code duplication"
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[lib]
proc-macro = true

[dependencies]
derive-syn-parse.workspace = true
proc-macro2.workspace = true
quote.workspace = true
syn.workspace = true
indexmap.workspace = true

[dev-dependencies]
eyre.workspace = true
trybuild.workspace = true

[package.metadata.release]
consolidate-commits = true
consolidate-pushes = true
dev-version = true
dev-version-ext = "dev"
post-release-commit-message = "v{{next_version}}"
pre-release-commit-message = "v{{version}}"
shared-version = true
tag-message = "https://crates.io/crates/{{crate_name}}/{{version}}"
tag-name = "{{version}}"
