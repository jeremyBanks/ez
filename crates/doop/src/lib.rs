pub(crate) mod evaluation;
pub(crate) mod input;
pub(crate) mod synthesis;

#[proc_macro]
pub fn doop(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let tokens: proc_macro2::TokenStream = tokens.into();

    let input: input::DoopBlock = match syn::parse2(tokens) {
        Ok(input) => input,
        Err(report) => return report.to_compile_error().into(),
    };

    let evaluation: evaluation::Doop = match input.try_into() {
        Ok(evaluation) => evaluation,
        Err(report) => return report.to_compile_error().into(),
    };

    let synthesis: synthesis::Doop = match evaluation.try_into() {
        Ok(synthesis) => synthesis,
        Err(report) => return report.to_compile_error().into(),
    };

    let tokens = synthesis.into_iter();
    let tokens: proc_macro2::TokenStream = tokens.collect();
    let tokens: proc_macro::TokenStream = tokens.into();

    tokens
}

#[proc_macro_attribute]
pub fn dooped(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let attr: proc_macro2::TokenStream = attr.into();
    let item: proc_macro2::TokenStream = item.into();

    let input = match input::DoopItem::try_from_tokens(attr, item) {
        Ok(input) => input,
        Err(report) => return report.to_compile_error().into(),
    };

    let evaluation: evaluation::Doop = match input.try_into() {
        Ok(evaluation) => evaluation,
        Err(report) => return report.to_compile_error().into(),
    };

    let synthesis: synthesis::Doop = match evaluation.try_into() {
        Ok(synthesis) => synthesis,
        Err(report) => return report.to_compile_error().into(),
    };

    let tokens = synthesis.into_iter();
    let tokens: proc_macro2::TokenStream = tokens.collect();
    let tokens: proc_macro::TokenStream = tokens.into();

    tokens
}
