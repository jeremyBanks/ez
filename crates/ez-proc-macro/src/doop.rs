#![allow(clippy::all)]
#![allow(dead_code)]

use {
    crate::common::{TokenTreeExt, TokenTreeIterExt},
    derive_syn_parse::Parse,
    indexmap::{IndexMap, IndexSet},
    proc_macro2::{Group, Ident, Span, TokenStream, TokenTree},
    quote::ToTokens,
    syn::{
        ext::IdentExt,
        parse::{Parse, ParseStream},
        punctuated::Punctuated,
        token, Token,
    },
};

#[derive(Debug, Clone)]
pub struct DoopBlock {
    pub items: Vec<DoopItem>,
}

impl Parse for DoopBlock {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut items = Vec::new();
        while !input.is_empty() {
            items.push(input.parse()?);
        }
        Ok(DoopBlock { items })
    }
}

#[derive(Parse, Debug, Clone)]
pub enum DoopItem {
    #[peek(Token![let], name = "let")]
    Let(DoopLetItem),
    #[peek(Token![for], name = "for")]
    For(DoopForItem),
}

#[derive(Parse, Debug, Clone)]
pub struct DoopForItem {
    #[call(DoopForBinding::parse_vec)]
    pub bindings: Vec<DoopForBinding>,
    #[brace]
    pub braces: token::Brace,
    #[inside(braces)]
    pub body: proc_macro2::TokenTree,
}

#[derive(Parse, Debug, Clone)]
pub struct DoopForBinding {
    #[prefix(Token![for])]
    pub target: ForBindingTarget,
    #[prefix(Token![in])]
    pub first_term: BindingTerm,
    #[call(RestTerm::parse_vec)]
    pub rest_terms: Vec<RestTerm>,
}

impl DoopForBinding {
    pub fn parse_vec(input: ParseStream) -> syn::Result<Vec<Self>> {
        let mut for_binding = Vec::new();
        while input.peek(Token![for]) {
            for_binding.push(input.parse()?);
        }
        Ok(for_binding)
    }
}

#[derive(Parse, Debug, Clone)]
pub enum ForBindingTarget {
    #[peek(syn::Ident::peek_any, name = "ident")]
    Ident(Ident),
    #[peek(token::Paren, name = "tuple")]
    Tuple(TupleBinding),
}

#[derive(Parse, Debug, Clone)]
pub struct TupleBinding {
    #[paren]
    pub paren: token::Paren,
    #[inside(paren)]
    #[call(Punctuated::parse_separated_nonempty)]
    pub items: Punctuated<Ident, Token![,]>,
}

#[derive(Parse, Debug, Clone)]
pub struct DoopLetItem {
    #[prefix(Token![let])]
    pub ident: Ident,
    #[prefix(Token![=])]
    pub first_term: BindingTerm,
    #[call(RestTerm::parse_vec)]
    pub rest_terms: Vec<RestTerm>,
    pub semi: Token![;],
}

#[derive(Parse, Debug, Clone)]
pub struct RestTerm {
    pub operation: PlusOrMinus,
    pub term: BindingTerm,
}

impl RestTerm {
    pub fn parse_vec(input: ParseStream) -> syn::Result<Vec<Self>> {
        let mut rest_terms = Vec::new();
        while input.peek(Token![+]) || input.peek(Token![-]) {
            rest_terms.push(input.parse()?);
        }
        Ok(rest_terms)
    }
}

#[derive(Parse, Debug, Clone)]
pub enum PlusOrMinus {
    #[peek(Token![+], name = "plus")]
    Plus(Token![+]),
    #[peek(Token![-], name = "minus")]
    Minus(Token![-]),
}

#[derive(Parse, Debug, Clone)]
pub enum BindingTerm {
    #[peek(syn::Ident, name = "ident")]
    Ident(Ident),
    #[peek(token::Bracket, name = "bracket list")]
    BracketedList(BracketList),
    #[peek(token::Paren, name = "paren list")]
    ParenList(ParenList),
    #[peek(token::Bracket, name = "brace list")]
    BraceList(BraceList),
}

#[derive(Parse, Debug, Clone)]
pub struct BracketList {
    #[bracket]
    bracket: token::Bracket,
    #[inside(bracket)]
    _todo: TokenStream,
}

#[derive(Parse, Debug, Clone)]
pub struct ParenList {
    #[paren]
    paren: token::Paren,
    #[inside(paren)]
    _todo: TokenStream,
}

#[derive(Parse, Debug, Clone)]
pub struct BraceList {
    #[brace]
    brace: token::Brace,
    #[inside(brace)]
    _todo: TokenStream,
}

pub fn doop(tokens: TokenStream) -> Result<TokenStream, eyre::Report> {
    let input: DoopBlock = syn::parse2(tokens)?;

    println!("{input:#?}");

    let mut output = TokenStream::new();

    let mut let_bindings = IndexMap::<Ident, IndexSet<TokenStream>>::new();

    Ok(output.into_iter().collect())
}

// We should also support punctuation subsitution?
fn replace_ident_in_token_stream(
    input: TokenStream,
    ident: &Ident,
    replacement: TokenStream,
) -> Result<TokenStream, syn::Error> {
    let mut output = TokenStream::new();
    for token in input {
        match token {
            TokenTree::Ident(ref candidate) =>
                if *candidate == *ident {
                    output.extend(replacement.clone().into_token_stream());
                } else {
                    output.extend([token.clone()]);
                },

            TokenTree::Group(group) => output.extend([TokenTree::Group(Group::new(
                group.delimiter(),
                replace_ident_in_token_stream(group.stream(), ident, replacement.clone())?,
            ))]),
            _ => output.extend([token.clone()]),
        }
    }
    Ok(output)
}
