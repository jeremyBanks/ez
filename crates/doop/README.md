`doop`—spelled like "loop" and pronounced like "dupe"—is a macro for local code duplication in Rust,
using a loop-style syntax.

This type of code duplication should usually be avoided in favour of more conventional language
features like traits (or even functions). However for certain niche cases, those might not be
suitable, and this may be the alternative you need.

## Example

```rust
use doop::doop;

doop! {}
```

### Expanded

```rust
```

## Alternatives

- [the `duplicate` crate](https://crates.io/crates/duplicate)
- [convoluted use of `macro_rules`](https://stackoverflow.com/q/37752133/1114)
