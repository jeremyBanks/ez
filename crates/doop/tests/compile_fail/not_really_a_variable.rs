use doop::doop;

fn main() {
    doop! {
        for N in [1, 2, 3] {
            let n = N;
            println!("{n}");
            println!("{N}");
        }
    }
}
