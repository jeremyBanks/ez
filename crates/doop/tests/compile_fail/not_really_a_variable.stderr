error[E0425]: cannot find value `N` in this scope
 --> tests/compile_fail/not_really_a_variable.rs:8:24
  |
8 |             println!("{N}");
  |                        ^ help: a local variable with a similar name exists: `n`
